_posts/en/2024-10-09-wallet-integration-guide.md

```diff
@@ -0,0 +1,590 @@
+or conflicting with another low-fee commitment transaction package that wouldn’t
+be promptly included in a block.
+
+To take the maximal advantage of this upgrade, wallets and backends should
+integrate with the **submitpackage** Bitcoin Core RPC command:
+
+```hack
+bitcoin-cli submitpackage ‘[“<commitment_tx_hex>”, “<anchor_spend_hex>”]’
+```
+
+Wallet implementations should integrate their software with the command using
+the commitment transaction as well as an anchor child spend to ensure inclusion
+into miners’ mempools with the appropriate feerate.
+
+Note: The RPC endpoint supports many-child, single-parent packages, but these
+will not propagate under the 1P1C relay update.
+
+After a sufficient number of nodes upgrade on the network, the LN protocol may
+be updated to drop the “update_fee” message, which has been a source of
+unnecessary force closes during fee spikes for years now. With removal of this
+protocol message, commitment transactions could be set to a static 1 sat/vbyte
+feerate. With TRUC transactions, we can ensure that competing commitment
+transactions with anchor spends are allowed to RBF each other over the network,
+and if there are competing output spends from the same commitment transaction,
+that RBF can occur no matter which output is being spent. TRUC transactions are
+also allowed to be 0-fee, allowing reduction in spec complexity. With TRUC’s
+sibling eviction, we can also drop the 1 block CSV locktimes, since we are no
+longer overly concerned with what unconfirmed outputs are being spent, as long
+as each party can spend a single output themselves.
+
+With TRUC + P2A Anchors, we can reduce the blockspace usage of the current two
+anchors down to a single keyless anchor. This anchor requires no commitment to a
+public key or  signatures, saving additional blockspace. The fee bumping can
+also be outsourced to other agents that have no privileged key material. Anchors
+could also consist of a single output with shared key material between the
+counterparties rather than P2A, at the cost of additional vbytes in the benign
+unilateral close case.
+
+Similar strategies can be pursued when implementing advanced features such as
+splicing, to reduce the risk of RBF pinning. For example, a TRUC channel splice
+that is less than 1kvB in size could CPFP another channel's unilateral close,
+without exposing the bumper to RBF pins. Subsequent bumps can be done in series
+by replacing just the channel splice transaction. This comes at the cost of
+revealing the TRUC transaction type during splices.
+
+As you can see, significant complexity can be avoided and savings achieved with
+the updated features, provided each transaction can fit into the 1P1C paradigm.
+
+### Ark
+
+Not all transaction patterns fit into the 1P1C paradigm. A prime example of this
+is [Ark][topic ark] outputs, which commit to a tree of presigned(or covenant committed)
+transactions to unroll a shared UTXO.
```

Comment:
I feel like this is missing a bit of background before you get into the ideal unilateral close, as a non-Ark-expert. Here are some questions that come up for me, and maybe they can be incorporated into the description here:
- When does a unilateral close happen? Is it initiated by a user, and the idea is they should be able to get their money out? Is it supposed to be a common occurrence?
- What transactions are involved in a unilateral close?
- Are these txns with shared UTXOs + presigned like in LN? Who can pin us - is it just the service provider, or can somebody else add a descendant to make it difficult to CPFP?
- Is there a time window? What happens if the unilateral close takes too long? Does the user lose money?

Reply:
> When does a unilateral close happen? Is it initiated by a user, and the idea is they should be able to get their money out? Is it supposed to be a common occurrence?

It happens primarily if the ASP(Ark Service Provider) goes offline, or is censoring users. It's used in the same sense as LN, to make it non-custodial. It should Never Happen(TM) for scalability reasons.

> What transactions are involved in a unilateral close?

A Merkle branch of the tree essentially, with log(n) inner nodes and a final state transaction. 

> Are these txns with shared UTXOs + presigned like in LN? 

Yes, starting with a shared root with all N participants(could be hundreds, thousands potentially).

>  Who can pin us - is it just the service provider, or can somebody else add a descendant to make it difficult to CPFP?

Either could. You can imagine thousands of people trying to unilaterally exit at the same time could be problematic with intentional and accidental pins. Each participant is attempting to get into a block, each posting their merkle branch, hitting chain limits pretty fast potentially if not carefully designed.

> Is there a time window?

There's a weeks(?) time window where the Ark "times out" and the ASP can simply sweep all the remaining utxos from the old tree, without having to unroll the tree.

Reply:
Punting to @instagibbs on this suggestion

Reply:
could add some text:

"If an Ark Service Provider (ASP) goes offline or process a transaction, the user can choose to do a unilateral exit, which involves the user submitting a series of transactions to unroll their branch place in the transaction tree. This requires O(logn) transactions. Difficulty can arise if other clients are also attempting to leave the tree exceeding mempool chain limits, or creating conflicting transactions with insufficient fees for timely inclusion in a block. If a particularly long time window elapses without inclusion, the ASP is able to recoup all the funds unilaterally, resulting in user funds loss."

Reply:
That's what I was looking for, thanks! Lmk if this is correct / can be added as penultimate sentence:
"Since all participants' transactions stem from the same root transaction, any user can pin another user's exit through dominating the descendant limit."

Reply:
Yes. Once the root is confirmed, then the population is split in half and this "game" repeats. For simplicity I think just sticking with what you said is best though.